### ISSUE TITLE
Dgraph server panics when shutdown

### ISSUE THREAD

**Issue by peterstace** [dgraph-io/dgraph#1939]
I saw the following panic when using Ctrl-C:

```
^C2017/12/20 13:07:38 run.go:336: Caught Ctrl-C. Terminating now (this may take a few seconds)...
2017/12/20 13:07:38 admin.go:61: Got clean exit request
2017/12/20 13:07:38 Stopped taking more http(s) requests. Err: accept tcp [::]:8080: use of closed network connection
2017/12/20 13:07:38 gRpc server stopped : accept tcp [::]:9080: use of closed network connection
2017/12/20 13:07:38 All http(s) requests finished.
2017/12/20 13:07:43 wal.go:118: Writing snapshot to WAL, metadata: {ConfState:{Nodes:[1] XXX_unrecognized:[]} Index:177 Term:2 XXX_unrecognized:[]}, len(data): 27
panic: send on closed channel

goroutine 182 [running]:
github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*DB).sendToWriteCh(0xc4200be300, 0xc43960a728, 0x1, 0x1, 0x17, 0x20, 0xc4359248e0)
	/home/petsta/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/db.go:609 +0x15f
github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*DB).batchSet(0xc4200be300, 0xc43960a728, 0x1, 0x1, 0x0, 0xc43960a728)
	/home/petsta/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/db.go:679 +0x4d
github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*DB).purgeVersionsBelow(0xc4200be300, 0xc433840a20, 0xc4359248e0, 0x17, 0x20, 0x47, 0x0, 0x0)
	/home/petsta/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/db.go:931 +0x453
github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger.(*ManagedDB).PurgeVersionsBelow(0xc42008c0c0, 0xc4359248e0, 0x17, 0x20, 0x47, 0x0, 0x0)
	/home/petsta/go/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/badger/managed_db.go:79 +0x127
github.com/dgraph-io/dgraph/posting.periodicPurgeOldVersions.func1()
	/home/petsta/go/src/github.com/dgraph-io/dgraph/posting/lists.go:244 +0x2a2
github.com/dgraph-io/dgraph/posting.periodicPurgeOldVersions()
	/home/petsta/go/src/github.com/dgraph-io/dgraph/posting/lists.go:254 +0x40
created by github.com/dgraph-io/dgraph/posting.Init
	/home/petsta/go/src/github.com/dgraph-io/dgraph/posting/lists.go:211 +0xe1
```

**Comment by peterstace** [comment]
This is possibly the same shutdown as seen in https://github.com/dgraph-io/dgraph/issues/1938

This isn't a duplicate though -- that ticket is about an unrelated problem.

**Comment by peterstace** [comment]
Looks like any write operation to badger will cause a panic if has already been closed. Simply opening, closing, then writing to badger causes the same problem in `sendToWriteCh`.

Badger should ideally give an error rather than panicking.

This seems more like a badger issue to me, so closing.

**Comment by pawanrawal** [comment]
I think this is caused because we don't have a mechanism for periodicPurgeOldVersions(which runs in a goroutine) to complete before shutdown. We have discussed this earlier and the idea was that the user of Badger shouldn't do writes after it has been closed.

**Comment by peterstace** [comment]
Ok, that sounds like a reasonable approach to me. I'll re-open this ticket and close the other one.

**Comment by janardhan1993** [comment]
https://github.com/dgraph-io/badger/issues/367, after this issue purge won't eat much cpu so we don't need that background goroutine which tries to purge slowly over time.
We can do aggresively during lru eviction without worrying about cpu usage

**Comment by pawanrawal** [comment]
I have merged a fix to master with d06e674ede4c23e8d7ea8bb87a1cffbe6a4f6bc1.
